pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -166,46 +189,218 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+				value: node.Labels[constraint.TopologyKey],
+			}
+			// addTopologyPairWithoutPods is a non-op if other pods match this pair
+			nodeTopologyMaps.addTopologyPairWithoutPods(pair)
+		}
+
+		appendTopologyPairsMaps(nodeTopologyMaps)
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+
+	if err := errCh.ReceiveError(); err != nil {
+		return nil, err
+	}
+
+	// calculate min match for each topology pair
+	topologyPairsPodSpreadMap.topologyKeyToMinPodsMap = make(map[string]int32, len(constraints))
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[constraint.TopologyKey] = math.MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit all portions of <topologyKey: any value>
+		// if we see 0 as min match of the topologyKey
+		if l := int32(len(podSet)); l < topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key] {
+			topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key] = l
+		}
+	}
+	return topologyPairsPodSpreadMap, nil
+}
+
+func getHardTopologySpreadConstraints(pod *v1.Pod) (constraints []v1.TopologySpreadConstraint) {
+	if pod != nil {
+		for _, constraint := range pod.Spec.TopologySpreadConstraints {
+			if constraint.WhenUnsatisfiable == v1.DoNotSchedule {
+				constraints = append(constraints, constraint)
+			}
+		}
+	}
+	return
+}
+
+func podMatchesAllSpreadConstraints(pod *v1.Pod, ns string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 || pod.Namespace != ns {
+		return false, nil
+	}
+	return podLabelsMatchesSpreadConstraints(pod.Labels, constraints)
+}
+
+// some corner cases:
+// 1. podLabels = nil, constraint.LabelSelector = nil => returns false
+// 2. podLabels = nil => returns false
+// 3. constraint.LabelSelector = nil => returns false
+func podLabelsMatchesSpreadConstraints(podLabels map[string]string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 {
+		return false, nil
+	}
+	podLabelSet := labels.Set(podLabels)
+	for _, constraint := range constraints {
+		selector, err := metav1.LabelSelectorAsSelector(constraint.LabelSelector)
+		if err != nil {
+			return false, err
+		}
+		if !selector.Matches(podLabelSet) {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
 // returns a pointer to a new topologyPairsMaps
 func newTopologyPairsMaps() *topologyPairsMaps {
 	return &topologyPairsMaps{topologyPairToPods: make(map[topologyPair]podSet),
 		podToTopologyPairs: make(map[string]topologyPairSet)}
 }
 
-func (topologyPairsMaps *topologyPairsMaps) addTopologyPair(pair topologyPair, pod *v1.Pod) {
+func (m *topologyPairsMaps) addTopologyPair(pair topologyPair, pod *v1.Pod) {
 	podFullName := schedutil.GetPodFullName(pod)
-	if topologyPairsMaps.topologyPairToPods[pair] == nil {
-		topologyPairsMaps.topologyPairToPods[pair] = make(map[*v1.Pod]struct{})
+	if m.topologyPairToPods[pair] == nil {
```

Comment:
nit: Could we replace this we `addTopologyPairWithoutPods`?

Reply:
Done.